File: Greedy Algorithm/fractional_knapsack.py
def fractional_knapsack(profit, weight, capacity):
    # Return maximum value of items and their fractional amounts.

    # index = [0, 1, 2, ..., n - 1] for n items
    index = list(range(len(profit)))

    # contains ratios of profit to weight
    ratio = [v / w for v, w in zip(profit, weight)]

    # index is sorted according to profit-to-weight ratio in decreasing order
    index.sort(key=lambda i: ratio[i], reverse=True)

    max_profit = 0

    fractions = [0] * len(profit)

    for i in index:
        if weight[i] <= capacity:
            fractions[i] = 1
            max_profit = max_profit + profit[i]
            capacity = capacity - weight[i]
        else:
            fractions[i] = capacity / weight[i]
            max_profit = max_profit + (profit[i] * capacity / weight[i])
            break

    return max_profit, fractions



if __name__ == '__main__':

    num = int(input('Enter number of items : '))

    weight = []
    profit = []

    print('Enter the weight of each items :')
    for i in range(num):
        a = float(input())
        weight.append(a)

    print('Enter the profit of each items :')
    for i in range(num):
        a = float(input())
        profit.append(a)

    capacity = int(input('\nEnter maximum weight : '))

    max_value, fractions = fractional_knapsack(profit, weight, capacity)

    print('\nThe maximum profit value of items that can be carried :', max_value)
    print('\nThe fractions in which the items should be taken :', fractions)

File: Greedy Algorithm/fractional_knapsack.c
#include <stdio.h>


void knapsack(int num, float weight[], float profit[], float capacity)
{
    float x[20];
    float total_profit = 0;
    float cap = capacity;

    int i;
    for (i=0; i<num; i++) {
        x[i] = 0;
    }

    for (i=0; i<num; i++)
    {
        if(weight[i] > cap)
            break;
        else
        {
            x[i]=1;
            total_profit = total_profit + profit[i];
            cap = cap - weight[i];
        }
    }

    if(i < num)
        x[i] = cap/weight[i];

    total_profit = total_profit + (x[i] * profit[i]);

    printf("\nThe weight of each object taken is :\n");
    for(i=0; i<num; i++) {
        printf("%.2f\t", x[i]);
    }

    printf("\n\n Maximum profit is : %.2f", total_profit);
}



int main() {
    float weight[20];
    float profit[20];
    float capacity;
    int num;
    float ratio[20];


    printf("\nEnter the number of objects : ");
    scanf("%d", &num);

    printf ("\nEnter the weight of each object :\n");
    for (int i=0; i<num; i++)
    {
        scanf("%f", &weight[i]);
    }

    printf ("\nEnter the profit of each object :\n");
    for (int i=0; i<num; i++)
    {
        scanf("%f", &profit[i]);
    }

    printf ("\nEnter the capacity of knapsack : ");
    scanf ("%f", &capacity);

    printf("\nProfit per weight of all the objects are :\n");
    for(int i= 0; i<num; i++){
        ratio[i] = profit[i] / weight[i];
        printf("%.2f\n", ratio[i]);
    }

    // sorting the objects in descending order on the basis of array ratio[]
    // selection sort
    for(int i=0; i<num; i++)
    {
        float temp;
        for(int j=i+1; j < num; j++)
        {
            if(ratio[i]<ratio[j])
            {
                // swapping ratio
                temp= ratio[j];
                ratio[j]= ratio[i];
                ratio[i]= temp;

                // swapping weight
                temp= weight[j];
                weight[j]= weight[i];
                weight[i]= temp;

                // swapping profit
                temp= profit[j];
                profit[j]= profit[i];
                profit[i]= temp;
            }
        }
    }

    knapsack(num, weight, profit, capacity);

    return 0;
}
